### ISSUE TITLE
Integers sometimes have an extra 0 appended in v0.6.0

### ISSUE THREAD

**Issue by FletcherCuttingMachShip** [cue-lang/cue#2649]
### What version of CUE are you using (`cue version`)?

<pre>
v0.6.0
</pre>

### Does this issue reproduce with the latest stable release?
No it doesn't

### What did you do?
I updated cue to v0.6.0, ran my test suite and saw that in some tests integers has an extra 0 appended to the end of it.

I have isolated the issue into two tests:

```go
func Test_FloatAverage(t *testing.T) {
	cuectx := cuecontext.New()

	value := cuectx.CompileString(`
		import "list"
		list.Avg([4, 8, 12])
	`)

	if err := value.Err(); err != nil {
		t.Fatalf("failed to compile cue value: %s", err)
	}

	f64, err := value.Float64()

	if err != nil {
		t.Fatalf("failed to get float64 from cue value: %s", err)
	}

	t.Logf("float value: %v", f64)
}

func Test_IntAverage(t *testing.T) {
	cuectx := cuecontext.New()

	value := cuectx.CompileString(`
		import "list"
		list.Avg([4, 8, 12])
	`)

	if err := value.Err(); err != nil {
		t.Fatalf("failed to compile cue value: %s", err)
	}

	i64, err := value.Int64()

	if err != nil {
		t.Fatalf("failed to get int64 from cue value: %s", err)
	}

	t.Logf("int value: %v", i64)
}
```

### What did you expect to see?
I expected the following logs:
`Test_FloatAverage`:
```
float value: 8
```

`Test_IntAverage`:
```
int value: 8
```

### What did you see instead?
I received the following logs:
`Test_FloatAverage`:
```
float value: 8
```

`Test_IntAverage`:
```
int value: 80
```

### Using `Int()`
When running the following test:
```go
func Test_IntAverageBigInt(t *testing.T) {
	cuectx := cuecontext.New()

	value := cuectx.CompileString(`
		import "list"
		list.Avg([4, 8, 12])
	`)

	if err := value.Err(); err != nil {
		t.Fatalf("failed to compile cue value: %s", err)
	}

	i64, err := value.Int(nil)

	if err != nil {
		t.Fatalf("failed to get int64 from cue value: %s", err)
	}

	t.Logf("int value: %v", i64)
}
```
I get a panic with the following information (sensitive information redacted):
```
panic: cue: exponent should always be nil for integer types [recovered]
        panic: cue: exponent should always be nil for integer types

goroutine 146 [running]:
testing.tRunner.func1.2({0x10111d520, 0x1011e4ea0})
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1526 +0x1c8
testing.tRunner.func1()
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1529 +0x384
panic({0x10111d520, 0x1011e4ea0})
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:884 +0x204
cuelang.org/go/cue.Value.Int({0x1400019b638?, 0x140001f6480?, 0x0?}, 0x0)
       ***
        *** +0xa8
testing.tRunner(0x14000003860, 0x1011e3a88)
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1576 +0x10c
created by testing.(*T).Run
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/testing/testing.go:1629 +0x368
FAIL    *** 0.485s
```
I haven't dug deeper into what could be causing this issue, I assume it has something to do with converting from a `float` to an `int` under the hood and somehow a 0 is being appended to the end.

**Comment by mvdan** [comment]
Not having looked at this bug in detail yet, it sounds worrying. Have you been able to git bisect when the regression happened? That would be really helpful to understand the root cause. If not possible, knowing what CUE version worked for you, or for this particular example, would also help.

**Comment by amilicms** [comment]
Hello,
For me the last working stable CUE version was v0.5.0. `git bisect` reports [this commit](https://github.com/cue-lang/cue/commit/d573e0c2f0419f3a32e9d7bc1f3000cc365a240d) as the first bad commit.

I ran the following git bisect commands. I don't have a lot of experience using `git bisect` so please let me know if I've done something incorrectly.
```sh
$ git bisect start
$ git bisect bad v0.7.0
$ git bisect good d573e0c2f0419f3a32e9d7bc1f3000cc365a240d
$ git bisect run go run ./testint
```

Note: `d573e0c2f0419f3a32e9d7bc1f3000cc365a240d` is the first common ancestor of `v0.7.0` and `v0.5.0`, and my test program also returns zero for this commit.

Here is the full output of `git bisect run go run ./testint`:
```
running  'go' 'run' './testint'
Expected: 8, actual: 80
Bisecting: 103 revisions left to test after this (roughly 7 steps)
[f5277b924d3658135f37d211024b8bcac6767330] internal/ci: remove hard-codings in gerrithub package
running  'go' 'run' './testint'
Bisecting: 51 revisions left to test after this (roughly 6 steps)
[d5dd9ec2fd17c3584806cc71c9b2633cd1276226] internal/core/export: always export required fields
running  'go' 'run' './testint'
Bisecting: 25 revisions left to test after this (roughly 5 steps)
[49a64f4c69a3fe3310a78126feb1600f6d46ed36] internal/core/convert: split the cue tag commas after the last quote
running  'go' 'run' './testint'
Bisecting: 12 revisions left to test after this (roughly 4 steps)
[8868ae6fdf931de36973640b4bbc5fca5fc2722e] all: go fix to remove old build tags
running  'go' 'run' './testint'
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[b2476bd08a39851483e8dfdb27808f1b4cd60d9b] encoding/protobuf/jsonpb: remove unused code
running  'go' 'run' './testint'
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[cb461fd4741dcd9e621e0a53f1d2fe7b72c276bf] internal/cuetest: remove support for CUE_LONG
running  'go' 'run' './testint'
Bisecting: 1 revision left to test after this (roughly 1 step)
[12908b32a9a74c40ee14cf69ab7221a94bbdafd3] deps: switch to use quicktest instead of testify
running  'go' 'run' './testint'
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[f3be10a2114c5a97bdde29266cea8dc88e0daf0e] cue: add a "decimal" benchmark
running  'go' 'run' './testint'
47282232b4368f36621f6e4a3b0445bcc42afb9a is the first bad commit
commit 47282232b4368f36621f6e4a3b0445bcc42afb9a
Author: Daniel Martí <mvdan@mvdan.cc>
Date:   Thu Nov 3 15:42:04 2022 +0000

    update to cockroachdb/apd v3
    
    The main breaking change that apd/v3 brings is that it no longer uses
    math/big types directly. There are new wrapper types which avoid
    allocations in many common cases. See the added calls to MathBigInt and
    SetMathBigInt.
    
    The other breaking change happened in Quo and Sqrt,
    causing the operations in CUE to end up with different precision:
    
            === RUN   TestMarshalJSON
            === RUN   TestMarshalJSON/22/a:_1.0/1
                    types_test.go:2758:
                             got {"a":1.00000000000000000000000};
                            want {"a":1.0}
            --- FAIL: TestMarshalJSON (0.00s)
                    --- FAIL: TestMarshalJSON/22/a:_1.0/1 (0.00s)
    
    Get the old behavior back by overwriting the Quo and Sqrt methods,
    running a short piece of code to do what v2 used to do.
    This makes the unit tests pass again.
    
    As a small bonus, whenever we use Quo or Rem on an apd float,
    the resulting Decimal is now consistently reduced,
    leaving at least one digit after the decimal point.
    
    For example, cue/testdata/resolve/006_arithmetic.txtar worked out
    "4.0 / 2.0" as "2", but now it is "2.0".
    Similarly, cue/testdata/basicrewrite/002_arithmetic.txtar worked out
    "1.0T / 2.0" as "5.0000000000E+11", but now it is "5.0E+11".
    
    To properly change the behavior of Quo and Rem across the codebase,
    internal.BaseContext is now of type internal.Context,
    which embeds apd.Context. internal.BaseContext is now used everywhere,
    except for cue/literal, as that would result in cyclic imports.
    For now that is fine, since that package uses neither Quo nor Rem.
    
    If we want cue/literal to use the same custom apd.Context as the rest of
    the codebase, we could always move internal.Context to avoid the cycle.
    
    Performance is largely unchanged; there are likely more gains to be made
    if we reduce the amount of times we convert to math/big types.
    
        goos: linux
        goarch: amd64
        pkg: cuelang.org/go/cue/testdata/benchmarks
        cpu: AMD Ryzen 7 PRO 5850U with Radeon Graphics
                         │  apd-v2-24  │             apd-v3-24             │
                         │   sec/op    │   sec/op     vs base              │
        /decimal.txtar-8   4.982m ± 1%   5.045m ± 0%  +1.27% (p=0.004 n=6)
    
                         │  apd-v2-24   │             apd-v3-24              │
                         │     B/op     │     B/op      vs base              │
        /decimal.txtar-8   1.873Mi ± 0%   1.872Mi ± 0%  -0.09% (p=0.002 n=6)
    
                         │  apd-v2-24  │             apd-v3-24             │
                         │  allocs/op  │  allocs/op   vs base              │
        /decimal.txtar-8   29.72k ± 0%   29.65k ± 0%  -0.22% (p=0.002 n=6)
    
    Fixes #1872.
    
    Signed-off-by: Daniel Martí <mvdan@mvdan.cc>
    Change-Id: Iadcb4a50f7c9943601a1124570387f62ed65a6f5
    Reviewed-on: https://review.gerrithub.io/c/cue-lang/cue/+/543421
    TryBot-Result: CUEcueckoo <cueckoo+gerrithub@cuelang.org>
    Reviewed-by: Marcel van Lohuizen <mpvl@gmail.com>
    Unity-Result: CUE porcuepine <cue.porcuepine@gmail.com>

 cue/literal/num.go                             |  3 +-
 cue/path.go                                    |  2 +-
 cue/testdata/basicrewrite/002_arithmetic.txtar |  4 +-
 cue/testdata/benchmarks/decimal.txtar          |  2 +-
 cue/testdata/resolve/006_arithmetic.txtar      |  8 ++--
 cue/types.go                                   | 15 ++++---
 encoding/openapi/types.go                      |  2 +-
 encoding/protobuf/jsonpb/decoder.go            |  2 +-
 go.mod                                         |  4 +-
 go.sum                                         | 10 ++---
 internal/core/adt/context.go                   |  2 +-
 internal/core/adt/decimal.go                   | 36 +++++++----------
 internal/core/adt/expr.go                      |  2 +-
 internal/core/adt/simplify.go                  | 16 ++++----
 internal/core/compile/label.go                 |  2 +-
 internal/core/convert/go.go                    | 12 ++++--
 internal/core/convert/go_test.go               |  2 +-
 internal/core/export/bounds.go                 |  2 +-
 internal/internal.go                           | 54 ++++++++++++++++++++++++--
 internal/pkg/context.go                        |  2 +-
 pkg/list/math.go                               |  2 +-
 pkg/math/manual.go                             | 21 +++++-----
 pkg/math/math.go                               | 26 ++++++-------
 23 files changed, 135 insertions(+), 96 deletions(-)
bisect found first bad commit
```

Here's the source of testint:
```go
package main

import (
	"fmt"
	"os"

	"cuelang.org/go/cue/cuecontext"
)

// Exit code of:
// * 0 indicates the test passed and we got the value we expect
// * 1 indicates the test failed and we got the bogus value
// * 2 indicates there was some other problem compiling the cue or retrieving the value
func main() {
	cuectx := cuecontext.New()
	v := cuectx.CompileString(`
		import "list"
		list.Avg([4, 8, 12])
	`)

	if err := v.Err(); err != nil {
		fmt.Printf("Failed to compile cue value: %s\n", err)
		os.Exit(2)
	}

	i64, err := v.Int64()

	if err != nil {
		fmt.Printf("Failed to retrieve value as int64: %s\n", err)
		os.Exit(2)
	}

	if i64 != 8 {
		fmt.Printf("Expected: 8, actual: %d\n", i64)
		os.Exit(1)
	}
}
```

**Comment by myitcv** [comment]
@amilicms thanks for the detailed follow-up here and reproducer. As you will see from https://cuelang.org/cl/1173262, we have now fixed this. There is likely a bit more improvement/cleanup in terms of our code base as far false expectations on apd are concerned, but the fix just submitted should be generally correct.
